frontend/controller/actions/chatroom.js

```diff
@@ -255,43 +254,58 @@ export default (sbp('sbp/selectors/register', {
+    await sbp('chelonia/contract/retain', userIDs, { ephemeral: true })
+    try {
+      await sbp('chelonia/contract/wait', params.contractID)
 
-    if (!userID || !has(rootState.contracts, userID)) {
-      throw new Error(`Unable to send gi.actions/chatroom/join on ${params.contractID} because user ID contract ${userID} is missing`)
-    }
+      userIDs.forEach(cID => {
+        if (!cID || !has(rootState.contracts, cID) || !has(rootState, cID)) {
+          throw new Error(`Unable to send gi.actions/chatroom/join on ${params.contractID} because user ID contract ${cID} is missing`)
+        }
+      })
 
-    const CEKid = params.encryptionKeyId || await sbp('chelonia/contract/currentKeyIdByName', params.contractID, 'cek')
+      const CEKid = params.encryptionKeyId || await sbp('chelonia/contract/currentKeyIdByName', params.contractID, 'cek')
 
-    const userCSKid = sbp('chelonia/contract/currentKeyIdByName', userID, 'csk')
-    return await sbp('chelonia/out/atomic', {
-      ...params,
-      contractName: 'gi.contracts/chatroom',
-      data: [
+      const userCSKids = await Promise.all(userIDs.map(async (cID) =>
+        [cID, await sbp('chelonia/contract/currentKeyIdByName', cID, 'csk')]
+      ))
+      return await sbp('chelonia/out/atomic', {
+        ...params,
+        contractName: 'gi.contracts/chatroom',
+        data: [
         // Add the user's CSK to the contract
-        [
-          'chelonia/out/keyAdd', {
+          [
+            'chelonia/out/keyAdd', {
             // TODO: Find a way to have this wrapping be done by Chelonia directly
-            data: [encryptedOutgoingData(params.contractID, CEKid, {
-              foreignKey: `sp:${encodeURIComponent(userID)}?keyName=${encodeURIComponent('csk')}`,
-              id: userCSKid,
-              data: rootState[userID]._vm.authorizedKeys[userCSKid].data,
-              permissions: [GIMessage.OP_ACTION_ENCRYPTED + '#inner'],
-              allowedActions: '*',
-              purpose: ['sig'],
-              ringLevel: Number.MAX_SAFE_INTEGER,
-              name: `${userID}/${userCSKid}`
-            })]
-          }
+              data: userCSKids.map(([cID, cskID]: [string, string]) => encryptedOutgoingData(params.contractID, CEKid, {
+                foreignKey: `shelter:${encodeURIComponent(cID)}?keyName=${encodeURIComponent('csk')}`,
+                id: cskID,
+                data: rootState[cID]._vm.authorizedKeys[cskID].data,
+                permissions: [GIMessage.OP_ACTION_ENCRYPTED + '#inner'],
+                allowedActions: '*',
+                purpose: ['sig'],
+                ringLevel: Number.MAX_SAFE_INTEGER,
+                name: `${cID}/${cskID}`
+              }))
+            }
+          ],
+          ...(Array.isArray(userID)
+            ? userID.map(cID => sendMessage({ ...params, data: { memberID: cID }, returnInvocation: true }))
+            : [sendMessage({ ...params, returnInvocation: true })]
+          )
```

Comment:
This is pretty cool

Reply:
BTW, this is more or less why I was using `undefined` for ourselves.

Reply:
AFAICT this code wouldn't need to be changed

Reply:
It would because we generally use the convention that we don't set our own member ID. I don't think things _break_ if we do, but it makes for shorter messages and it's less likely to accidentally exclude us from the chatroom.

Reply:
Using `identityContractID` now